app/services/onlinetesting/Phase2TestService.scala

```diff
@@ -95,22 +96,12 @@ trait Phase2TestService extends OnlineTestService with Phase2TestConcern with Sc
     }
   }
 
-  def verifyAccessCode(email: String, accessCode: String): Future[Option[String]] = {
+  def verifyAccessCode(email: String, accessCode: String): Future[Try[String]] = {
```

Comment:
I think @benjstephenson has a point. Future has the concept of Failure baked in, and putting a Try in it just makes the caller work harder. If this were a private method it would not be such an issue, but muddying a public API because of the internal implementation is really a last resort. Perhaps using Future.fromTry would enable you to use your for comprehension over futures to simplify. Something like:
```
  def verifyAccessCode(email: String, accessCode: String): Future[String] = {
    for {
      userId <- cdRepository.findUserIdByEmail(email)
      testGroupOpt <- phase2TestRepo.getTestGroupByUserId(userId)
      value <- Future.fromTry(processEtrayToken(testGroupOpt, accessCode))
    } yield value
  }
```